src/server/common.cc

```diff
@@ -278,16 +282,43 @@ void Context::Cancel() {
 
 void Context::Reset(ErrHandler handler) {
   std::lock_guard lk{mu_};
+  CheckHandlerFb();
   err_ = {};
   err_handler_ = std::move(handler);
   Cancellation::flag_.store(false, std::memory_order_relaxed);
 }
 
 GenericError Context::Switch(ErrHandler handler) {
   std::lock_guard lk{mu_};
-  if (!err_)
+  if (!err_) {
+    // No need to check for the error handler - it can't be running
+    // if no error is set.
     err_handler_ = std::move(handler);
+  }
   return err_;
 }
 
+void Context::Stop() {
+  CheckHandlerFb();
+}
+
+GenericError Context::ReportInternal(GenericError&& err) {
+  std::lock_guard lk{mu_};
+  if (err_)
+    return err_;
+  err_ = std::move(err);
+
+  CheckHandlerFb();
```

Comment:
`CheckHandlerFb` -> `WaitForErrorHandlerToFinish()` ? Or a more concise version of it? 😄